contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,248 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
+* **RunAsUser**: UID to run the entrypoint of the container process. After usernamespace support in kubelet, if Kubelet discovers that usernamespace remapping is enabled at runtime, this will be the UID in the container usernamespace. In such a case, if no mapping(corresponding host UID) is found in the mapping range, pod will be failed admission at Kubelet. If kubelet discovers that usernamespace remapping is not supported at runtime, no validation check will be performed. 
+* **RunAsGroup**: same as **RunAsUser** but for gid mapping ranges.
+* **RunAsNonRoot**: Today, kubelet fails to start the container if image is found to be having UID 0. After usernamespace support in kubelet, kubelet should ignore this check if runtime is found to be running with usernamespace remapping enabled. Kubelet should verify image UID only when it is learned that usernamespace remapping is either not supported or not enabled at runtime.
+* **FsGroup and SupplementalGroups**: FsGroup and SupplementalGroups will continue to work after usernamespace support is Kubelet if runtime either does not support usernamespaces or support is not enabled. In case usernamespace support is discovered by kubelet as enabled at runtime, FsGroups/SupplementalGroups will work only if supported by the runtime (CRI implementation). One implementation, for example, could be where runtime(CRI implementation) will add a triplet in `/proc/<container-pid>/gid_map` for the FsGroup/Supplemental GIDs. This will be a 1:1 mapping between host usernamespace group id and container usernamespace group id i.e container ID and host ID will be same. Example:
 
-- **Beta:** Explicit API to request host user-namespace in pod spec
-          This is being targeted under Beta graduation plans.
+    If pod has `FsGroup: 2000` and runtime's default gid mappings are `containerid 0,  hostid 100000, size 1000`, runtime should be adding two entries in the `/proc/<container-pid>/gid_map`:
+    - [0 100000 1000]  // From GID 0 to 1000 in the container are mapped with GID range 100k to 100k+1000 on host
+    - [2000 2000 1]    // GID 2000 in container usernamespace is mapped to GID 2000 is host usernamespace
 
-### CRI API Changes
-Proposed CRI API changes:
+    Once this 1:1 mapping is set for the container process, `runc` already has the logic to call `setgroups()` to set the FsGroupID as the supplemental ID in the container usernamespace.
 
-```golang
-// Runtime service defines the public APIs for remote container runtimes
-service RuntimeService {
+    Setting the permissions on volume is resposibility of volume plugin. Once GID on the volume matches the supplemental GID in the container, access will work.
+
+    NOTE: Pod will be failed admission at kubelet if FsGroup/SupplementalGroups are in the default mapping range and mappings are not 1:1 i.e host id for a container id is different.
```

Comment:
Soft failure (`BLOCKED`) or hard failure? It seems problematic that this failure isn't known until the pod is actually run on a node.

Reply:
should be hard failure because pod spec or runtime configuration will not change on retries and does not seem recoverable. WDYT?

Reply:
The problem with hard failure is that the controllers will retry and cause churn. This is a general problem though, which we've discussed fixing in the controller layer.